### PR TITLE
Deprecate JsonElement constructor

### PR DESCRIPTION
**Pull request by Marcono1234**
Follow-up for #1689.

Creating custom JsonElement subclasses is discouraged and will lead to issues when using JsonTreeReader.

### CODE REVIEW

**Comment by Marcono1234** [comment]
@eamonnmcmanus, could you please have a look?

**Comment by eamonnmcmanus** [comment]
Thanks! I agree with the principle. I had a couple of small comments.
Another approach that would avoid having so many `@SuppressWarnings("deprecation")` annotations would be to have a second, package-private `JsonElement` constructor with a dummy parameter, and use that in the subclasses. But I think `@SuppressWarnings` is probably better.

**Comment by eamonnmcmanus** [comment]
Experimentally, I tried just changing the `JsonElement` constructor to package-private and testing that against all of Google's internal code. There were no failures. So, even though it is technically an incompatible API change, I think maybe we could allow ourselves to make it. Clearly the existing default public constructor was an oversight.

**Comment by Marcono1234** [comment]
> So, even though it is technically an incompatible API change, I think maybe we could allow ourselves to make it. Clearly the existing default public constructor was an oversight.

To be safe, I think keeping the constructor for now might be better because at least the Auto.js project has a [custom subclass](https://github.com/hyb1996/Auto.js/blob/468c46f61e74482a7a62b9737adda351cb7c3a77/app/src/main/java/org/autojs/autojs/tool/SafeJsonElement.java). Also, unlike some other issues, keeping the deprecated `JsonElement` constructor would probably not be such a big problem; and we can still remove it in the future.

File: gson/src/test/java/com/google/gson/internal/bind/JsonTreeReaderTest.java [hunk 1]
```diff
@@ -54,4 +56,28 @@ public void testHasNext_endOfDocument() throws IOException {
...
+    array.add(new CustomSubclass());
+
+    JsonTreeReader reader = new JsonTreeReader(array);
+    reader.beginArray();
+    try {
+      // Should fail due to custom JsonElement subclass
+      reader.peek();
+      fail();
+    } catch (MalformedJsonException expected) {
+      assertEquals("Custom JsonElement subclass com.google.gson.internal.bind.JsonTreeReaderTest$1CustomSubclass is not supported",
```
[the remark this answers is not in the document]
Comment (Marcono1234) [reply gson/src/test/java/com/google/gson/internal/bind/JsonTreeReaderTest.java:79]: Good point, that hardcoded generated name `JsonTreeReaderTest$1CustomSubclass` would likely cause problems in the future.

**Comment by eamonnmcmanus** [comment]
OK, given that we know of one subclass I agree that the change here is the right way to go.

**Pull request merged**